src/tuner/nccl_ofi_model.c

```diff
@@ -0,0 +1,717 @@
+}
+
+static void model_get_coll_info_internal_v2_p5_p5e(nccl_ofi_tuner_model_context_t *model_ctx, ncclFunc_t collType,
+						   size_t nBytes, int nvlsSupport, int numPipeOps, int *algorithm,
+						   int *protocol, int* nChannels)
+{
+	float cost = 0;
+	float lowest = FLT_MAX;
+	int algo, proto = 0;
+
+	/* Skip runs smaller than 2 nodes and fallback to NCCL's internal tunings */
+	if (model_ctx->dims.num_nodes <= 2) {
+		return;
+	}
+
+	if (collType == ncclFuncAllReduce && model_ctx->dims.num_nodes == 16 &&
+	    model_ctx->dims.num_ranks == 128 && nvlsSupport && nBytes > 3ULL * 1024ULL * 1024ULL * 1024ULL &&
+	    nBytes <= 5ULL * 1024ULL * 1024ULL * 1024ULL) {
+		lowest = 0;
+		*algorithm = NCCL_ALGO_NVLS_TREE;
+		*protocol = NCCL_PROTO_SIMPLE;
+		goto exit;
+	}
+
+	/*
+	 * Ideally, this should just be a lookup and not be in-flight math
+	 * We do not want divs in the hot path, but working with the API we've
+	 * got now.
+	 */
+	for (algo = 0; algo < NCCL_NUM_ALGORITHMS; algo++) {
+		/* No CollNet on AWS today */
+		if (algo == NCCL_ALGO_COLLNET_DIRECT || algo == NCCL_ALGO_COLLNET_CHAIN)
+			continue;
+
+		/* Skip NCCL_ALGO_NVLS used only for single-node jobs */
+		if (algo == NCCL_ALGO_NVLS)
+			continue;
+
+		if (!nvlsSupport && (algo == NCCL_ALGO_NVLS_TREE))
+			continue;
+
+		for (proto = 0; proto < NCCL_NUM_PROTOCOLS; proto++) {
+			/* This is not a supported combination in NCCL */
+			if (algo == NCCL_ALGO_NVLS_TREE && proto != NCCL_PROTO_SIMPLE)
+				continue;
+
+			cost = nccl_ofi_tuner_compute_cost_p5_p5e(&model_ctx->model_params, &model_ctx->dims,
+								  collType, algo, proto, numPipeOps,  nBytes);
+			if (cost < 0)
+				continue;
+
+			NCCL_OFI_TRACE(NCCL_TUNING, "Model Tuner Computed cost for algo %d proto %d pipe %d: cost %.8f µsecs.",
+				       algo, proto, numPipeOps, cost);
+			if (cost < lowest) {
+				*algorithm = algo;
+				*protocol = proto;
+				lowest = cost;
+			}
+		}
+	}
+
+exit:
+	NCCL_OFI_INFO(NCCL_TUNING, "Model Tuner Choosing algo %d proto %d with cost %.8f µsecs for coll %d size %ld.",
+				    *algorithm, *protocol, lowest, collType, nBytes);
+}
+
+static ncclResult_t model_init_internal_p5_p5e(nccl_ofi_tuner_model_context_t *model_ctx)
+{
+	int func;
+	int algo, proto = 0;
+
+	/*
+	 * P5/P5e platform has 4 NIC rails per GPU. Each rail with 100Gbps bandwidth
+	 */
+
+	/*
+	 * For Hopper GPUs on P5, all intranode communication goes over NVLink, so use
+	 * the bandwidth for SM90 architecture in NCCL (SM90_NVLINK_BW).
+	 *
+	 * This is unidirectional bandwidth per NVLink (900GB/s bidirectional on the
+	 * platform, with 18 NVLinks in total. NCCL considers a 20% protocol overhead,
+	 * leaving 20GB/s bandwidth per link).
+	 */
+	const struct nccl_ofi_tuner_model_params params_p5_p5e = {
+		.net_lat = 20.0,
+		.internode_bw = (12.5 * 1024 * 1024 * 1024 * 1e-6) /* per rail */,
+		.intranode_bw = (20.0 * 1024 * 1024 * 1024 * 1e-6),
+		.num_rails = 4 
+	};
```

Comment:
Same comment as above, we should grab this from platform_data values instead.

Reply:
yes, that's possible, but only if model logic can be common, and latency and bandwidth are parameters for different platforms. As I answered above, this will limit the flexibility of adding different model and manual selections per platform. For that reason, I prefer to contain tuner related parameters and code inside tuner code.